IP space visualization module

rasovica opened the pull request:
This module will add IP space map to nodewatcher, which will visualize and report statistics about network IP space usage.

rasovica committed:
Added cidr map tab to menu, added cidr module.

rasovica committed:
Added cidr module to settings.py.

rasovica committed:
Set the menu entry weight.

rasovica committed:
Added needed css file which has css for tooltips(author https://github.com/jaz303) and a few extra classes that make graph non selecteble.

rasovica committed:
Fixed mixed indentation.

rasovica committed:
Fixed mixed indentation.

rasovica committed:
Merge branch 'feature/cidr' of https://github.com/wlanslovenija/nodewatcher into feature/cidr

rasovica committed:
Fixed another hidden tab.

mitar commented:
In which way is this different from #58?

rasovica committed:
Refractor cidr module to ip_space for better understanding.

rasovica committed:
Fixed settings.py to match new module name.

rasovica committed:
Refractored css, refractored title, added needed javascript files to static folder, Mike Bostock is author of d3.v4, hilbert2d.js and point.js are refractored from node.js version to browser version by me, original author(https://github.com/ryan-williams), jquery.tips.js is made by https://github.com/jaz303 this modules provide all i need to draw svg maps with tooltips using hilbert fractals.

rasovica committed:
Moved javascript and css to local static folder.

kostko commented on nodewatcher/settings.py:
```diff
@@ -665,6 +666,11 @@ def user_url(user):
             'visible': True,
         },
+        {
+            'name': 'Ip space',
```
Rename to "IP space".

rasovica replied:
Done

kostko commented on nodewatcher/modules/importer/nw2/management/commands/import_nw2.py:
```diff
@@ -277,7 +277,7 @@ def import_pools(self, data):
                 family='ipv4',
                 network=pool['network'],
-                prefix_length=pool['cidr'],
+                prefix_length=pool['ip_space'],
```
These changes in `import_nw2.py` seem incorrect - was this a too broad search and replace?

robimarko replied:
That should not be touched from what I understand.
That is only used to import old Nodewatcher 2 JSON dump for migration

rasovica replied:
Fixed

kostko replied:
It doesn't look fixed, there are still some incorrect `ip_space` references in `import_nw2.py`? See diff.

robimarko commented:
Please add some description so most people can at least get a sense of what you are doing here

mitar commented:
Please put code attribution into code comments, not commit messages.

mitar commented:
Don't use minimized JavaScript files. If we need those, we will just have minimization post-processing step for all files.

rasovica committed:
Renamed to IP space, removed unused html files, fixed javascript export error.

rasovica committed:
Replaced minified d3.v4 with source, changed the file in template.

rasovica committed:
Fixed accidental serach and replace mistake.

rasovica committed:
Added authors of files into comments.

rasovica committed:
Added topnode calculation, zooming to topnodes, drawing actual api nodes to svg, loading nodes asyncronusly and drawaing them to svg.

kostko commented on nodewatcher/modules/frontend/ip_space/frontend.py:
```diff
@@ -0,0 +1,34 @@
+
+components.partials.get_partial('ip_space_partial').add(components.PartialEntry(
+    name='general',
+    template='ip_space.html',
```
Is this ok that the template is at the top level? I think it should be better to move it under `network/ip_space/ip_space.html` or similar.

kostko replied:
I also don't see where these partials are currently used? Are they needed at all? If not, they should be removed.

kostko replied:
@rasovica where are you using these partials?

kostko commented:
BTW, please fix your GitHub or the commits' e-mail address as they don't seem to correspond (unknown author).

rasovica committed:
Rounding down map size to closes power of two in an attempt to fix the positioning bug.

rasovica committed:
Adding zoom and pan to ip space overview.

rasovica committed:
Fixing the miss-alignment of networks.

kostko commented on nodewatcher/settings.py:
```diff
@@ -665,6 +666,11 @@ def user_url(user):
         },
+        {
+            'name': 'IP space',
+            'weight': 70,
```
Should this be placed before the API entry instead?

rasovica replied:
I would put it before Map entry so it's next to network toplogoy

kostko replied:
Agreed.

kostko commented on nodewatcher/modules/importer/nw2/management/commands/import_nw2.py:
```diff
@@ -386,12 +386,12 @@ def import_nodes(self, data):
 
             try:
-                translated_subnets = SUBNET_SIZE_TRANSLATION[subnet_mesh['cidr']]
+                translated_subnets = SUBNET_SIZE_TRANSLATION[subnet_mesh['ip_space']]
```
This is still incorrect.

kostko commented on nodewatcher/modules/importer/nw2/management/commands/import_nw2.py:
```diff
@@ -386,12 +386,12 @@ def import_nodes(self, data):
 
             # Allocate Router-ID based on subnet translation
-            subnet_mesh['cidr'] = translated_subnets['rid']
+            subnet_mesh['ip_space'] = translated_subnets['rid']
```
This is still incorrect.

kostko commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space.html:
```diff
@@ -0,0 +1,50 @@
+        </div>
+    </div>
+    <script>
+    $(document).ready(function(){
```
Can you put this into a separate Javascript file and not inline it? You can check how it is done in other modules.

rasovica committed:
Fixed the remaining replace errors.

rasovica committed:
Moved all javascript to files.

rasovica committed:
Updating the menu position.

rasovica committed:
Removed unneeded partials.

rasovica committed:
The module now uses local api and not the nodes.wlan-si.net one which was used for testing.

rasovica committed:
Fixing blank line for pep8 compliance.

rasovica committed:
Merge branch 'development' into feature/cidr

rasovica committed:
Merge branch 'development' into feature/cidr

kostko commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+    {% add_data "js_data_end" "js/DrawCidrs.js" %}
+    {% add_data "js_data_end" "js/main.js" %}
+
+
```
Remove empty line.

rasovica replied:
Done

kostko commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+    {% add_data "js_data" "js/point.js" %}
+    {% add_data "js_data" "js/hilbert2d.js" %}
+    {% add_data "js_data_end" "js/jquery.tipsy.js" %}
+    {% add_data "js_data_end" "js/DrawCidrs.js" %}
```
Can you rename `DrawCidrs.js` to something like `draw_cidrs.js`? We don't want camel case in filenames.

rasovica replied:
Done

kostko commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+    <div class="row">
+        <div class="col-md-2">
+            <div class="dropdown">
+                <button class="btn btn-primary dropdown-toggle" type="button" data-toggle="dropdown">Zoom to subnet<span class="caret"></span></button>
```
Make strings translatable.

rasovica replied:
Done

kostko commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+            <div class="dropdown">
+                <button class="btn btn-primary dropdown-toggle" type="button" data-toggle="dropdown">Zoom to subnet<span class="caret"></span></button>
+                <ul class="dropdown-menu" id="topnodes">
+                    <li id="cidr"><a href="#" value="10.0.0.0/8">Loading data</option>
```
Make strings translatable.

rasovica replied:
And done.

kostko commented on nodewatcher/settings.py:
```diff
@@ -728,4 +735,4 @@ def user_url(user):
 # Allowed hosts (required for production use, when DEBUG is false). Set it
 # to the (virtual HTTP) hostname under which you have nodewatcher installation.
-ALLOWED_HOSTS = []
+ALLOWED_HOSTS = ["0.0.0.0"]
```
Why this change?

rasovica replied:
Mistake, will revert.

kostko reviewed (commented):
Did another review pass, only some minor comments, otherwise looks good. @mitar can you also do a pass?

rasovica committed:
Removing mistake.

rasovica committed:
Refractor to draw_cidrs.

rasovica committed:
Removing empty line.

rasovica committed:
Adding translatable strings.

rasovica committed:
Merge branch 'feature/cidr' of https://github.com/wlanslovenija/nodewatcher into feature/cidr

mitar commented on nodewatcher/core/frontend/static/frontend/css/style.scss:
```diff
@@ -11,3 +11,4 @@ $bootstrap-sass-asset-helper: true;
 @import "menu";
 @import "node";
 @import "charts";
+
```
Why this change?

mitar reviewed (changes requested):
Please fix your commit information so that it is something GitHub can recognize and attribute to you.

Feel free to git force push to your branch rewritten commits.

mitar commented on nodewatcher/modules/frontend/ip_space/static/css/ipspace.css:
```diff
@@ -0,0 +1,42 @@
+    width: 100%;
+    transform-origin: top left;
+}
+/*Author of tipsy: https://github.com/jaz303 */
```
What is tipsy and where it did come from? Is there an upstream repository?

mitar replied:
Can this just be included, instead of  copied?

rasovica replied:
I included repository information, and moved tipsy css to own file, however this is not exact copy of the tipsy repository as it is modified to work correctly with the d3.js.

mitar commented on nodewatcher/modules/frontend/ip_space/static/css/ipspace.css:
```diff
@@ -0,0 +1,42 @@
+    user-select: none; /* Non-prefixed version, currently
+                      supported by Chrome and Opera */
+}
+.silky-smooth{
```
Please make style unified. Make newlines between each block. Spaces before `{`.

rasovica replied:
Done.

kostko commented:
Oh, yes, you should also squash your commits as needed.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,147 @@
+window.DrawCidr = class DrawCidr{
```
Please also make unified style here. Two spaces for indentation. See our other JavaScript code for example of code style. In short: we use `;` at the end f lines. We use `'` for strings, except if user-facing strings. We put spaces around `{`. We always go with a new block into new line.

mitar commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,31 @@
+            </div>
+        </div>
+        <div class="col-md-8" id="wraper">
+            <svg id="ipspace" class="silky-smooth"></svg>
```
What is this class name? What does it mean?

rasovica replied:
It is a reference to hit movie You Don't Mess with the Zohan from 2008, it does all kinds of stuff for the main map window, do you want me to rename it to map or something like that?

mitar replied:
Very interesting. Yes, please make it something bring but related to the project.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/point.js:
```diff
@@ -0,0 +1,86 @@
+//Author: https://github.com/ryan-williams
```
So we want more than just author information. Where is code coming from?

rasovica replied:
Added repository it is based from.

mitar commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,31 @@
+            <div class="dropdown">
+                <button class="btn btn-primary dropdown-toggle" type="button" data-toggle="dropdown">{% trans "Zoom to subnet" %}<span class="caret"></span></button>
+                <ul class="dropdown-menu" id="topnodes">
+                    <li id="cidr"><a href="#" value="10.0.0.0/8">{% trans "Loading data" %}</option>
```
Don't use `#` in links, because they can destroy potential anchor information.

Why is this a link anyway?

rasovica replied:
Yes it was some weird html(closed </option> even tough there is no start), i fixed it now, it need to be a to be drawn correctly in the dropdown however.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/main.js:
```diff
@@ -0,0 +1,17 @@
+    var svgContainer = d3.select("#ipspace");
+    var size = svgContainer.node().getBoundingClientRect().width;
+    svgContainer.style("height", size + 'px');
+    var main_subnet = "10.0.0.0/8";
```
Why is this hardcoded? Maybe fetch this from template? Like you are doing for `#cidr`? So initialize from it.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/main.js:
```diff
@@ -0,0 +1,17 @@
+        alert(this.value);
+        var scale= size/cidr.subnetSize(this.value);
+        var start = cidr.subnetXY(this.value);
+        console.log(start);
```
No logging output please for final version.

mitar reviewed (changes requested):
One feature request. Currently when you are moving around in the space, there is no way that I can share the URL to a location in the space for somebody else to use it. Please make it so that you are changing URL anchor string in a way that one can then share it. If it would be a map, something like `#lat=...&lon=...&zoom=...` would be an example. Not sure what are here relevant values. Maybe cidr input box value, x and y?

rasovica committed:
Unified css style, moved tipsy css to its own file.

rasovica committed:
Unified javascript files.

rasovica committed:
Fixed weird html. Removed console and alert.

mitar commented:
Also, don't hard-code it to `10/8`, but maybe make it so that Django template renders what is the initial pool, and then you use that from JavaScript. See how we do [`data-source`](https://github.com/wlanslovenija/nodewatcher/blob/development/nodewatcher/modules/frontend/list/templates/nodes/list.html#L33) for example here.

kostko commented:
> Also, don't hard-code it to 10/8, but maybe make it so that Django template renders what is the initial pool, and then you use that from JavaScript. See how we do data-source for example here.

To choose the top-level range, you could query all the existing top-level pools in the database and use the shortest prefix, which contains all the given ranges.

rasovica committed:
Better code attribution.

rasovica committed:
No idea how this got here, removing empty line.

rasovica committed:
Rename css class to no fun name.

rasovica committed:
Fixing url to be non static.

rasovica committed:
Add url map sharing.

rasovica commented:
I added a feature mitar requested, if you zoom in the url sets to something like this `ip_space/?scale=45.25483399593904&start=191,3775` and if you load the site with this parameters set it zooms in acordingly 
![screenshot from 2017-09-02 07-02-27](https://user-images.githubusercontent.com/3155661/29993076-b95edd26-8fac-11e7-98c2-e7560446ecb6.png)

mitar commented:
What does `start` stand for in URL?

mitar commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+            <div class="dropdown">
+                <button class="btn btn-primary dropdown-toggle" type="button" data-toggle="dropdown">{% trans "Zoom to subnet" %}<span class="caret"></span></button>
+                <ul class="dropdown-menu" id="topnodes">
+                    <li id="cidr" value="10.0.0.0/8">{% trans "Loading data" %}</li>
```
This is still hard-coded?

mitar replied:
This is still hard-coded?

mitar commented on nodewatcher/modules/frontend/ip_space/templates/network/ip_space/ip_space.html:
```diff
@@ -0,0 +1,32 @@
+    {% add_data "js_data_end" "js/draw_cidrs.js" %}
+    {% add_data "js_data_end" "js/main.js" %}
+
+    {% trans "Ip space" as title %}
```
IP space.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/main.js:
```diff
@@ -0,0 +1,15 @@
+  var svgContainer = d3.select('#ipspace');
+  var size = svgContainer.node().getBoundingClientRect().width;
+  svgContainer.style('height', size + 'px');
+  var main_subnet = '10.0.0.0/8';
```
?

mitar replied:
This is still hard-coded?

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/point.js:
```diff
@@ -0,0 +1,86 @@
+//Author: https://github.com/ryan-williams, from: https://github.com/ryan-williams/hilbert-js
```
Please also made a comment with what changes you made to it, if any.

mitar commented on nodewatcher/modules/frontend/ip_space/static/css/ipspace.css:
```diff
@@ -0,0 +1,27 @@
+.noselect {
+    -webkit-touch-callout: none;
+    /* iOS Safari */
```
Why comments after? Put them before code.

mitar commented on nodewatcher/modules/frontend/ip_space/static/css/tipsy.css:
```diff
@@ -0,0 +1,125 @@
+/*Author of tipsy: https://github.com/jaz303, tipsy repository: https://github.com/jaz303/tipsy*/
```
Spaces inside comments. So between `/*` and text. Feel free to use multi-line comment for a header.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/hilbert2d.js:
```diff
@@ -0,0 +1,145 @@
+//Author: https://github.com/ryan-williams, from: https://github.com/ryan-williams/hilbert-js
```
Could you put all files from one external repository into one directory?

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/jquery.tipsy.js:
```diff
@@ -0,0 +1,284 @@
+//Author: https://github.com/jaz303
```
Where is this coming from? What are changes? Why not simply use  git submodule and pull that code in? Or make a fork of it and have changes there and then git submodule here?

rasovica committed:
Add zoom to fit button.

rasovica committed:
IP space as title.

rasovica committed:
Adding proper attribution to all js/css files, switching around css comments the right way around.

rasovica commented:
The start= is the IP in the furthest most left top corner converted to hilbert curve 2d in the x,y notation. The 10.0.0.0/8 is hardcoded while loading data, because we don't yet know what the network is gonna be like, I added last night a button zoom to fit
![screenshot from 2017-09-03 17-02-25](https://user-images.githubusercontent.com/3155661/30004128-c00615b8-90c9-11e7-8414-f28a34547033.png)
Which calculates the smallest possible network which fits all the top-level nodes, however I don't zoom to it automatically, just list it as an option to zoom to, I don't think zooming to it instantly is a good idea, after all we want users to see how much of the 10.0.0.0/8 is left and where the top nodes are in that space, If they have a small network they can zoom to fit it on the screen by simply pressing one button.

mitar commented:
> The 10.0.0.0/8 is hardcoded while loading data, because we don't yet know what the network is gonna be like
So you compute the smallest network on the server side, render it to the template, and then use that to initialize client side. So you do not have to compute that on the client.

> I don't think zooming to it instantly is a good idea, after all we want users to see how much of the 10.0.0.0/8 is left and where the top nodes are in that space

10.0.0.0/8 is not special in any way. Why then don't start with 0.0.0.0/0? And show how much of the whole Internet is left? :-)

No, please do so that it starts with the smallest network which includes all IP pools nodewatcher instance has. Because user cares only about the free space in IP pools. Not free space of 10.0.0.0/8.

Alternatively, you could render one of your visualizations per non-nesting IP pools.

rasovica committed:
Map now starts fitted to the network. Also fixed a function returning undefined.

rasovica commented:
The network is now fitted to the map on start.

mitar commented on nodewatcher/modules/frontend/ip_space/static/css/tipsy.css:
```diff
@@ -0,0 +1,129 @@
+/* Author of tipsy: https://github.com/jaz303
+   tipsy repository: https://github.com/jaz303/tipsy
+   modified by: https://bl.ocks.org/ilyabo
+   copied from: http://bl.ocks.org/ilyabo/1373263
```
Please create a `wlanslovenija` branch in this repo (from `master` or `v2` branch, whichever is closer to http://bl.ocks.org/ilyabo/1373263): https://github.com/wlanslovenija/tipsy

Then, in that branch, apply changes from http://bl.ocks.org/ilyabo/1373263.

Then, use git submodule to pull that branch in here.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/d3.v4.js:
```diff
@@ -0,0 +1,16866 @@
+// https://d3js.org Version 4.9.1. Copyright 2017 Mike Bostock.
```
Use git submodule to pull this in.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/hilbert2d.js:
```diff
@@ -0,0 +1,148 @@
+/* Author: https://github.com/ryan-williams
+   from: https://github.com/ryan-williams/hilbert-js
+   changes: This used to be a require/node module, I adapted it for in browser module, removed 3d hilbert curve, removed unneeded hilbert.js which was used for init, as I only need one module and not two.
```
Do the same here: https://github.com/wlanslovenija/hilbert-js

Create a branch `wlanslovenija`, and add all changes there. Then use git submodule to pull that code in.

mitar commented:
If d3 is not changed in any way, then you can just add a static file as you have done. But if anything is changed, then do it through a git submodule.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,275 @@
+                    self.draw(data.results[j].network + "/" + data.results[j].prefix_length, data.results[j].description);
+                }
+                self.data.push(data);
+                if (self.data.length == 33) {
```
What is this? What is this hard-coded value?

rasovica committed:
Fixing wrong loading of the data.

rasovica committed:
Adding git repository as submodules for tipsy and hilbert-js, leaving the d3.v4 static.

rasovica commented:
I added submodules, symlink to static files. Just waiting for the premission to make branch the commits are ready on my side. The hardcoded number of 33 is the number of all possible prefixses from 0-32, however i found a bug/ bad design in my code where i would assume i will only get one page of data and not follow the next url, i have now implemented an recursive function that loads the data, for wlan-si the full page load now takes about 100s for about 7000nodes.

mitar commented:
> I have now implemented an recursive function that loads the data, for wlan-si the full page load now takes about 100s for about 7000nodes.

Hm, didn't we initially said that you will render things lazily? So you render just top level things, then recurse when user zooms and so on? But probably it is hard to know when one can do that?

Why do you fetch things in order of prefix length? Why don't you just fetch everything in one request? You are fetching at the end everything, no?

rasovica committed:
Switching submodules to correct branch.

rasovica committed:
Removing console from final version.

rasovica committed:
Fixing browser overflow, this is needed because the initial map is huge and get scaled down (so we are not working with floating point pixels, but whole pixels) and this leaves the browser side bars even after scale-down.

rasovica committed:
Removing the test link.

mitar commented on .gitmodules:
```diff
@@ -0,0 +1,8 @@
+[submodule "libs/hilbert-js"]
+	path = libs/hilbert-js
+	url = https://github.com/wlanslovenija/hilbert-js
+    branch = wlanslovenija
```
You added this manually? Indentation does not mach.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+        mask = 32-mask;
+        var start_ip = this.num2ip(this.offset + start_num);
+
+        $('#topnodes').html('<li id="cidr" cidr="10.0.0.0/8"><a>10.0.0.0/8 (Zoom out)</a></li>');
```
Why is `10.0.0.0/8` here? There is no zoom out to 10.0.0.0/8. 10.0.0.0/8 is just arbitrary. Zoom out always to smallest network which fits.

mitar replied:
Why is `10.0.0.0/8` still here?

mitar replied:
You are recreating whole topnodes?

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+
+        var ips_needed = stop_num - start_num;
+        var mask = 0;
+        while(Math.pow(2, mask) < ips_needed){
```
Space between `while` and `(`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+        var max_x = self.getMaxOfArray(xs);
+        var max_y = self.getMaxOfArray(ys);
+
+
```
No unnecessary empty line.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+    }
+
+    loadPrefix(i, url){
+            var self = this;
```
Wrong identation.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+                    for (var j = 0; j < data.results.length; j++) {
+                        self.draw(data.results[j].network + "/" + data.results[j].prefix_length, data.results[j].description);
+                    }
+                }else{
```
How with `else` into a new line.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+                    self.loadPrefix(i, data.next);
+                    if(self.data[i] == undefined){
+                        self.data[i] = data.results;
+                    }else{
```
Put `else` on its own line.

rasovica committed:
Fixing indentetion.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+                        self.data[i].concat(data.results);
+                    }
+                    for (var j = 0; j < data.results.length; j++) {
+                        self.draw(data.results[j].network + "/" + data.results[j].prefix_length, data.results[j].description);
```
Use `'` and not `"` for non-user facing strings.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,297 @@
+        this.svg.append('rect').attr('x', x).attr('y', y).attr('height', shape_size).attr('width', shape_size).style('fill', this.rgbToHex(times * 8, times * 8, times * 8)).style('opacity', 0.3).attr('id', subnet).attr('n', subnet_ips).attr('d', description);
+    }
+
+    loadPrefix(i, url){
```
Change to `loadPrefix(i, url) {`.

rasovica committed:
Fixing indentation and formatting.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+            hash;
+
+        if (re.test(url)) {
+            if (typeof value !== 'undefined' && value !== null)
```
Always use `{`, do not have this ifs without.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+
+    UpdateQueryString(key, value, url) {
+        if (!url) url = window.location.href;
+        var re = new RegExp("([?&])" + key + "=.*?(&|#|$)(.*)", "gi"),
```
`'` for non-user facing strings.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+                    self.draw(data.results[j].network + '/' + data.results[j].prefix_length, data.results[j].description);
+                }
+            }
+            else{
```
Space before `{`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+            }
+            else{
+                self.loadedPrefixes++;
+                if(self.data[i] == undefined){
```
Space around `()`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+                self.loadedPrefixes++;
+                if(self.data[i] == undefined){
+                    self.data[i] = data.results;
+                }else{
```
Into new line.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+                    jQuery('rect').tipsy({
+                        gravity: 'w',
+                        html: true,
+                        title: function() {
```
For anonymous function, use `function () {` style.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+
+    load() {
+        for (var i = 0; i < 33; i++) {
+            this.loadPrefix(i, 'api/v2/pool/ip/?format=json&prefix_length=' + i);
```
Get URL API from server-side and render-it into a template. See examples elsewhere how to do that.

mitar replied:
You do not have to add `format=json` because `getJSON` adds needed headers already.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+       range[0] = (this.ip2num(cidr[0])) & ((-1 << (32 - cidr_1)));
+       var start = this.ip2num(range[0])
+       range[1] = range[0] + Math.pow(2, (32 - cidr_1)) - 1;
+       return [range[0]-this.offset, range[1]-this.offset];
```
Space around operators.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,298 @@
+        while( Math.pow(2, mask) < ips_needed){
+            mask++;
+        }
+        mask = 32-mask;
```
Space around operators.

rasovica committed:
Fixing a lot of stuff, from formatting to how minimal network is calculated.

rasovica committed:
Removing console.

rasovica committed:
Adding urltemplate support.

rasovica committed:
Fixing blank lines.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        this.svg = svg;
+        this.loadedPrefixes = 0;
+
+        this.n_ips = Math.pow(2, 32 - parseInt(subnet.split("/")[1]));
```
`'` instead `"`. Please when I make a comment, make sure yourself to check everywhere else for the same issue.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        this.ips_side = Math.sqrt(this.n_ips);
+        this.ips_pixel = this.ips_side / (this.size * 1.0);
+        this.offset = 0;
+        this.draw(subnet, "Smallest possible network to fit all the ip pools");
```
`ip` -> `IP`

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        this.data = new Array(33);
+        this.base_url = url;
+        self = this;
+        $('#topnodes').on('click', '#cidr', function(event) {
```
You are already using `class`, so you can then also use `(event) => { ... }` which will bind `this` automatically so you do not have to use `self`.

mitar replied:
BTW, style for anonymous functions is `function (event) {`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        return false;
+    }
+
+    UpdateQueryString(key, value, url) {
```
Camel case. Start with lower case.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+
+    subnetSize(subnet) {
+        var subnet_str = String(subnet);
+        var times = 32 - parseInt(subnet_str.split("/")[1]);
```
`'`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+    HSVtoRGB(h, s, v) {
+        var r, g, b, i, f, p, q, t;
+        if (arguments.length === 1) {
+            s = h.s, v = h.v, h = h.h;
```
Split this into multiple lines. No `,`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        t = v * (1 - (1 - f) * s);
+        switch (i % 6) {
+            case 0:
+                r = v, g = t, b = p;
```
Same.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        return '#' + this.componentToHex(r) + this.componentToHex(g) + this.componentToHex(b);
+    }
+
+    get_url_param(sParam) {
```
Use `camelCase` not `under_scores`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+
+    get_url_param(sParam) {
+        var sPageURL = decodeURIComponent(window.location.search.substring(1)),
+            sURLVariables = sPageURL.split('&'),
```
No `,`. Simply define one var per line.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+    }
+
+    UpdateQueryString(key, value, url) {
+        if (!url){
```
Space after `)`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+            else {
+                hash = url.split('#');
+                url = hash[0].replace(re, '$1$3').replace(/(&|\?)$/, '');
+                if (typeof hash[1] !== 'undefined' && hash[1] !== null)
```
No `if` without `{...}`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+                hash = url.split('#');
+                url = hash[0] + separator + key + '=' + value;
+                if (typeof hash[1] !== 'undefined' && hash[1] !== null)
+                    url += '#' + hash[1];
```
Same.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+                    url += '#' + hash[1];
+                return url;
+            }
+            else
```
Same.

rasovica committed:
Removing last console.log.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+
+    loadPrefix(i, url) {
+        var self = this;
+        $.getJSON(url, function(data) {
```
You can use `(data) => {...}`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+        $.getJSON(url, function(data) {
+            if(data.next != null){
+                self.loadPrefix(i, data.next);
+                if (self.data[i] == undefined) {
```
Always use `===`.

mitar commented on nodewatcher/modules/frontend/ip_space/static/js/draw_cidrs.js:
```diff
@@ -0,0 +1,265 @@
+    loadPrefix(i, url) {
+        var self = this;
+        $.getJSON(url, function(data) {
+            if(data.next != null){
```
Spaces. Always use `!==`.

rasovica committed:
Fixing pep8 and removing some unused functions.

rasovica committed:
Adding Support module. Adding comments, replacing functions with does from Support module.

rasovica committed:
Removed unneeded libraries, updated template to replace word node with pool, removed git modules that we don't need.

rasovica committed:
Adding recursive utilization calculation.
